### PR TITLE
Distributed channel - config-kafka reloading

### LINKED ISSUE #455: Remove on-the-fly reloading of `config-kafka` in distributed channel
**Problem**
I started with on-the-fly reloading of `config-kafka` in consolidated channel: https://github.com/knative-sandbox/eventing-kafka/issues/318

Distributed channel already does this, but it is missing in consolidated channel.

I tried this for a while but in the end the code was getting too complex and also there were some technical issues (e.g. some stuff read from `config-kafka` can only be set once in the underlying libraries, otherwise memory leaks can happen). In the end, I decided to cancel this.

Since we're trying to align both channels, I think it is a good idea to also remove this reloading in distributed channel (their implementation also suffers from the problems above). @travis-minke-sap  agreed that we should remove this in there as well: https://groups.google.com/g/knative-dev/c/u1-P03-VZh0

**UPDATE:**
- Consolidated channel now reloads the configuration. However, dispatcher doesn't do on-the-fly reloading. Instead it restarts.
- https://github.com/knative-sandbox/eventing-kafka/pull/423: Watch `config-kafka` in controller
- https://github.com/knative-sandbox/eventing-kafka/pull/486: Do global resync of everyting if `config-kafka` is updated. This triggers reconciliation of the CRs and dispatcher deployments.
- https://github.com/knative-sandbox/eventing-kafka/pull/514: Trigger a rolling deployment of dispatcher in case the configmap is updated

We should do the same in distributed channel too.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Developer

**Exit Criteria**
No more on-the-fly reloading of `config-kafka` in distributed channel

**Time Estimate (optional):**
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Ticket for adding this in consolidated channel: https://github.com/knative-sandbox/eventing-kafka/issues/318
ML discussion: https://groups.google.com/g/knative-dev/c/u1-P03-VZh0

### PR DESCRIPTION
**Pull request by aliok**
Fixes #455

## Proposed Changes

1. Mount config-kafka to controller, dispatcher and receiver
2. Don't watch config-kafka in dispatcher and receiver
3. Make changes in the controller so that it does a redeployment of dispatcher and receiver when config-kafka changes

**Release Note**

```release-note
- 🧽  Distributed Kafka Channel dispatcher and receiver are now reconfigured when `config-kafka` configmap is updated

```

**Docs**

### CODE REVIEW

File: pkg/channel/distributed/common/kafka/sarama/sarama.go
Comment (eric-sap) [inline pkg/channel/distributed/common/kafka/sarama/sarama.go:44]: If we're going to have this configmap "loader" inside the context anyway, would it save some coding complexity (i.e. having to call configmaploader.FromContext(ctx) and the configmapLoader() call itself, including all of the error handling) to just have this LoadSettings function extract the configmaploader function from the context (which is already a parameter here) and call it, instead of calling that in each place that's calling LoadSettings() ?  It would involve fiddling with the tests a bit and I am not 100% sure the result would be "better" but thought I would mention it.

Comment (aliok) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go:44]: Sounds good. 
I have some changes stashed, but I will get back at this before merging the PR.

Comment (aliok) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go]: @eric-sap this is done now, thanks for the suggestion.

Comment (eric-sap) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go]: I like it; thanks!

Comment (aliok) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go:44]: I had to revert this later because I needed to load the config outside this function as well.
Specifically, when I needed to compute the configmap hash on controller start.

**Comment by aliok** [comment]
@eric-sap 
Pushed new commits that address the 3rd part in the PR description.
Sorry the PR got big now :(

I also did 1-2 things on the consolidated channel code in this PR, since I found some trivial problems when I compared the code with distributed channel.

**Comment by aliok** [comment]
/hold
I haven't done the manual tests yet

**Comment by aliok** [comment]
/unhold

I tested my changes with https://github.com/aliok/kafka-ch-reload-tests. There are 2 pods:

- sender sends messages for 2 mins
- receiver receives messages for 3 mins with a message frequency of 10 ms

There were no messages lost and no messages undelivered. Also no duplicate messages. I don't get how that's possible actually.

@eric-sap would that be expected? The channel would be unavailable for a few milliseconds in the consolidated channel when `config-kafka` is changed and thus sender can't send a few messages.
This doesn't happen here in distributed channel. 
Perhaps, we're reporting the consolidated dispatcher ready before it is actually ready.

@eric-sap I would use some help in manual testing if receiver deployment is updated when config-kafka is changed. Any instructions would be nice to make the reconcilers create a receiver deployment.

**Comment by eric-sap** [comment]
> @eric-sap would that be expected? The channel would be unavailable for a few milliseconds in the consolidated channel when `config-kafka` is changed and thus sender can't send a few messages.
> This doesn't happen here in distributed channel.

I haven't looked at all of the code here yet, but ideally when a deployment is changed, Kubernetes starts the new pod and does not switch the target or terminate the old pod until the readiness check says that the new pod is okay.  In the distributed channel's case, that means calling health.SetDispatcherReady() (in the dispatcher's main.go) and the SetChannelReady/SetProducerReady calls made after the receiver's components are created.  Theoretically this means that the instant the pod is "ready" it really can begin processing messages (and local strimzi testing has borne this out).

That said, I have seen duplicate messages happen during rollout restarts when using an external kafka broker and not local strimzi.  Our guarantee is only that no messages will be lost (that is, if a client receives a 202 from the receiver, that means the message will be delivered "at least one" (or retried until final failure and possible redelivery to a dead-letter-queue if configured).

File: pkg/channel/distributed/common/kafka/sarama/sarama.go
Comment (eric-sap) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go:44]: Doh; thanks for the effort!  Sorry for the extra time!

**Comment by aliok** [comment]
/hold
Receiver is not restarted. Having a look now

**Comment by aliok** [comment]
/unhold
Thanks @eric-sap.
I pushed the code change that fixes global resync in kafkaSecret controller

File: pkg/channel/distributed/controller/kafkachannel/dispatcher.go
Comment (eric-sap) [inline pkg/channel/distributed/controller/kafkachannel/dispatcher.go:308]: This is probably because of grabbing code from the consolidated side, but we already have the 'logger' variable available in this function, so re-extracting it from the context here(and a couple of other places; won't comment on them all) isn't necessary.

File: pkg/channel/distributed/controller/kafkasecret/controller.go
Comment (eric-sap) [inline pkg/channel/distributed/controller/kafkasecret/controller.go:102]: This comment doesn't match the code any more

File: pkg/channel/distributed/controller/kafkasecret/receiver.go
Comment (eric-sap) [inline pkg/channel/distributed/controller/kafkasecret/receiver.go:241]: Here also the logger variable is already available to use

File: pkg/channel/distributed/controller/kafkachannel/dispatcher.go
Comment (aliok) [reply pkg/channel/distributed/controller/kafkachannel/dispatcher.go]: Yeah, copy-paster error. Pushed a new commit now.

**Comment by eric-sap** [comment]
I've tested this locally and it seems to do what it's supposed to.  There are still some linter errors but other than that it looks good to me.

**Comment by aliok** [comment]
Linter error is about a deprecated dependency in test images. We need to fix it, but it is irrelevant to this PR.

@travis-minke-sap can you have a look and lgtm it?

**Review by travis-minke-sap — APPROVED** [review]
Thanks Ali !

**Pull request merged**
